`SpatialSeries` and `DynamicTraceSegment`

vhirtham opened the pull request:
## Changes

_Describe changes in this PR_

## Related Issues

Closes #691

## Checks

- [ ] updated CHANGELOG.rst
- [ ] updated tests
- [ ] updated doc/
- [ ] update example/tutorial notebooks
- [ ] update manifest file

CagtayFabry commented:
This looks really cool 👍 🚀 ! (and super fun to play with 😎 )

I know it's not final but I'll add some quick comments as notes in the code

CagtayFabry commented on weldx/geometry.py:
```diff
@@ -1577,15 +1578,73 @@ def local_coordinate_system(
+        self._max_s = max_s
+        self._length = self._len_expr() if series.is_expression else self._len_disc()
+
+    def _get_derivative(self, i):
+        me = self._series.data
+        exp = me.expression
+        # todo unit stripped -> how to proceed? how to cast all length units to mm?
+        subs = [(k, v[i].data.to_base_units().m) for k, v in me.parameters.items()]
+        return exp.subs(subs).diff("s")
```
for performance it would probably be good to calculate the derivative (and other formulas needed) once after instantiation

vhirtham replied:
There is actually a lot of room for improvements, especially in the "quality of life" section. But I just wanted to get things running first. :)

CagtayFabry commented on weldx/geometry.py:
```diff
@@ -1577,15 +1578,73 @@ def local_coordinate_system(
+        """Get the length of the segment."""
+        return self._length
+
+    def local_coordinate_system(self, position: float) -> tf.LocalCoordinateSystem:
+        """Calculate a local coordinate system at a position of the trace segment."""
+        if self._series.is_expression:
+            return self._lcs_expr(position)
+        return self._lcs_disc(position)
```
It would be great if we could pass an array for `position` to evaluate multiple `LCS` at the same time

vhirtham replied:
Yes. But for this to be useful we also need to rewrite the implementation of the `Trace`.

CagtayFabry commented on weldx/geometry.py:
```diff
@@ -1848,6 +1906,13 @@ def plot(
         else:
             axes.plot(data[0].m, data[1].m, data[2].m, fmt)
 
+    def _k3d_line(self, raster_width: pint.Quantity = "1mm"):
+        """Get (or show) a k3d line from of the trace."""
+        import k3d
+
+        r = self.rasterize(raster_width).to(_DEFAULT_LEN_UNIT).magnitude
+        return k3d.line(r.astype("float32").T)
```
I just added this for easier testing/debugging here

CagtayFabry commented on weldx/core.py:
```diff
@@ -1546,3 +1549,76 @@ def interp_like(
+            parameters = self._process_parameters(parameters)
+        super().__init__(obj, dims, coords, units, interpolation, parameters)
+
+    @staticmethod
+    def _process_quantity(
+        obj: Union[pint.Quantity, xr.DataArray, str, MathematicalExpression],
+        dims: Union[list[str], dict[str, str]],
+        coords: dict[str, pint.Quantity],
+    ) -> xr.DataArray:
+        """Turn a quantity into a a correctly formatted data array."""
+        if isinstance(coords, dict):
+            s = coords["s"]
+        else:
+            s = coords
+            coords = dict(s=s)
+
+        if not isinstance(s, xr.DataArray):
+            if not isinstance(s, Q_):
+                s = Q_(s, "")
+            s = xr.DataArray(s, dims=["s"]).pint.dequantify()
+            coords["s"] = s
+
+        if "c" not in coords:
+            coords["c"] = ["x", "y", "z"]
+
+        if dims is None:
+            dims = ["s", "c"]
+
+        return xr.DataArray(obj, dims=dims, coords=coords)
+
+    @staticmethod
+    def _process_parameters(params):
+        """Turn quantity parameters into the correctly formatted data arrays."""
+        for k, v in params.items():
+            if isinstance(v, Q_) and v.size == 3:
+                params[k] = xr.DataArray(v, dims=["c"], coords=dict(c=["x", "y", "z"]))
+        return params
```
can we use the `SeriesParameter` class here?
I think we would have to add a `coords` property to make this work

CagtayFabry commented on CHANGELOG.rst:
```diff
@@ -9,7 +9,7 @@
 =====
 
--  `SpatialSeries` and `DynamicTraceSegment` [:pull:`699`]
+-  ``SpatialSeries`` and ``DynamicTraceSegment`` [:pull:`699`]
```
have you tried `~weldx.core.SpatialSeries` ?
Maybe it works if you add it to the user API documentation in `weldx/__init__.py`

CagtayFabry replied:
I think because of the user API some elements (our main classes) appear two times in the sphinx tree, e.g. once as `weldx.time.Time` and once as `weldx.Time` (user API)
In these cases the references have to be explicit or sphinx might throw an error (because it doesn't know which page to link to). But that should be mostly on the API documentation site

For the changelog I am nur sure what the sphinx context is for resolving these (or if there is any difference)

CagtayFabry commented on weldx/geometry.py:
```diff
@@ -1379,70 +1381,241 @@ def shapes(self) -> list[Shape]:
+
+        self._length = self.get_section_length(self._max_s)
+
+    def _get_component_derivative_squared(self, i: int):
```
can you add the return value for the type hint here @vhirtham 
Is it a sympy expression?

vhirtham replied:
I will try it. When I printed the type, I got a derived type. I have to figure out the base type

CagtayFabry reviewed (approved):
Just some minor details, I have also modified the `GenericSeries` evaluation to work with correctly assigned array coordinates

We could probably change the `SeriesParameter` to always assign coordinates in `SeriesParameter.data_array` or are there places where we do not want that?

vhirtham commented:
great,... add a type hint,... get 9 unrelated warnings in the doc-build 👎

CagtayFabry commented:
> great,... add a type hint,... get 9 unrelated warnings in the doc-build 😕 👎

if unrelated just ignore and merge
We can just exclude `matplotlib` from the intersphinx mapping in `conf.py` since it is not really essential to `weldx` and the doc structure seems to cause a lot of issues

vhirtham commented:
Since we added a variable for the name of the dimension used by the `SpatialSeries`, I also used it in the `DynamicTraceSegment`. In case we do want to rename it for some reason, there shouldn't be any hardcoded dimension names in the segment class. I also renamed all the internal variables and parameters to be more general. @CagtayFabry Have another look at the changes and feel free to merge if everything is okay.

marscher commented on weldx/geometry.py:
```diff
@@ -1469,14 +1649,25 @@ def __init__(
         else:
             self._sign_winding = 1
 
+        # todo change sign sign back to + and correct winding signs?
```
Is this TODO still needed?

vhirtham replied:
depends... it basically just swaps the signs in the equation (so that it looks more like one would expect) and of the variable above. But in consequence, all tests have to be adjusted. Well, maybe I'll just try it and see how much work it is.
